server/src/main/java/org/elasticsearch/snapshots/InvalidSnapshotNameException.java

```diff
@@ -18,17 +22,63 @@
  */
 public class InvalidSnapshotNameException extends SnapshotException {
 
-    public InvalidSnapshotNameException(final String repositoryName, final String snapshotName, String desc) {
-        super(repositoryName, snapshotName, "Invalid snapshot name [" + snapshotName + "], " + desc);
+    @Nullable
+    private final Reason reason;
+
+    public InvalidSnapshotNameException(final String repositoryName, final String snapshotName, Reason reason) {
+        super(repositoryName, snapshotName, "Invalid snapshot name [" + snapshotName + "], " + reason.description);
+        this.reason = reason;
     }
 
     public InvalidSnapshotNameException(StreamInput in) throws IOException {
         super(in);
+        if (in.getVersion().onOrAfter(Version.V_8_2_0)) {
+            reason = in.readOptionalEnum(Reason.class);
```

Comment:
On 8.2.0 and above, will this really be optional? That is, maybe this should be `readEnum` and `writeEnum`?

Reply:
That's a good point, I can't think of a good reason to make it optional, I mean when deeming a name invalid the reasons are usually straightforward.... I will change it.